src/hardware/sblaster.cpp

```diff
@@ -2254,9 +2249,41 @@ class SBLASTER final {
 		sb.mixer.enabled=section->Get_bool("sbmixer");
 		sb.mixer.stereo=false;
 
-		sb.type = find_sbtype();
-		oplmode = find_oplmode();
 
+		// Determine Sound Blaster model
+		sb.type = [&] {
+			const std::string pref = section->Get_string("sbtype");
+
+			if (const auto maybe_sbtype = parse_sbtype_pref(pref);
+			    maybe_sbtype) {
+				return *maybe_sbtype;
+			} else {
+				LOG_WARNING("OPL: Invalid 'sbtype' setting: '%s', using 'sb16'",
+				            pref.c_str());
+				return SBT_16;
+			}
+		}();
+
+		// Determine OPL model
+		oplmode = [&] {
+			const std::string pref = section->Get_string("oplmode");
+
+			if (const auto maybe_oplmode = parse_oplmode_pref(pref, sb.type);
+			    maybe_oplmode) {
+				return *maybe_oplmode;
+			} else {
+				LOG_WARNING("OPL: Invalid 'oplmode' setting: '%s', using 'auto'",
+				            pref.c_str());
+
+				// Now it's guaranteed to succeed
+				const auto auto_oplmode = parse_oplmode_pref(pref,
```

Comment:
pref is unchanged, so it will actually fail again, unless I'm missing something. Do you intend to set "oplmode" to "auto" in this case, followed by a call to `parse_opl_mode("auto", sb.type)`?

Reply:
Uh... This actually works because of some really confusing and hacky interactions in `setup.cpp`. In the invalid case setting this path actually never runs... I'll leave tracing through it to you if you're _really_ interested; too complicated to explain 😄 

Anyway, you're absolutely right, if this _did_ work as one would logically hink, this would be a problem. Although this is dead code, fixing it as you suggested because I don't want to really on the hack, at least I want the correct implementation to be there. We will get rid of the legacy config handling some time and rewrite it properly, then this error handling logic will come in handy.